### PR TITLE
[ASAN] Instrument device global variables for OMP offloading

### PR DESCRIPTION
**Pull request by jinge90**

### CODE REVIEW

**Review by zhaomaosu — APPROVED** [review]
lgtm

File: llvm/lib/SYCLLowerIR/SanitizeDeviceGlobal.cpp [hunk 1]
```diff
@@ -43,12 +43,22 @@ static bool instrumentDeviceGlobal(Module &M) {
   //  size_t beginning address of the device global
   StructType *StructTy = StructType::get(IntTy, IntTy, IntTy);
 
+  const GlobalVariable *OMPOffloadEntry =
+      M.getNamedGlobal(".omp_offloading.entry_name");
```
Comment (abhinavgaba) [inline llvm/lib/SYCLLowerIR/SanitizeDeviceGlobal.cpp:47]: It's probably more reliable to use `M.getModuleFlag("openmp-device");` to check for OpenMP device compilation. See this for reference: https://github.com/intel/llvm/blob/2adf19646e48060f9d1a8fd143ed1276b4967c77/llvm/lib/Transforms/IPO/OpenMPOpt.cpp#L5944

Comment (jinge90) [reply llvm/lib/SYCLLowerIR/SanitizeDeviceGlobal.cpp]: Hi, @abhinavgaba 
Done.
Thanks very much!

**Review by abhinavgaba — APPROVED** [review]
Thanks, Ge.

**Comment by jinge90** [comment]
Hi, @intel/dpcpp-tools-reviewers 
Could you help review this patch?
Thanks very much.

File: llvm/lib/SYCLLowerIR/SanitizeDeviceGlobal.cpp [hunk 2]
```diff
@@ -43,12 +43,21 @@ static bool instrumentDeviceGlobal(Module &M) {
...
   for (auto &G : M.globals()) {
-    // Non image scope device globals are implemented by device USM, and the
-    // out-of-bounds check for them will be done by sanitizer USM part. So we
-    // exclude them here.
-    if (!isDeviceGlobalVariable(G) || !hasDeviceImageScopeProperty(G))
-      continue;
+    if (OMPDeviceMD) {
+      // For openmp offloading, the scope is to check external global variables
+      // within '#pragma declare target' block and ignore all internal names.
+      if (G.getName().starts_with("__") || G.getName().starts_with(".") ||
```
[the remark this answers is not in the document]
Comment (jinge90) [reply llvm/lib/SYCLLowerIR/SanitizeDeviceGlobal.cpp:51]: Hi, @sarnex 
For omp offloading, we only check device global variables from user, compiler will add some internal global variables whose name start with "." or "__", we exclude these 'internal' variables here.

Thanks very much.

File: llvm/lib/SYCLLowerIR/SanitizeDeviceGlobal.cpp [hunk 3]
```diff
@@ -43,12 +43,21 @@ static bool instrumentDeviceGlobal(Module &M) {
...
+          !G.hasExternalLinkage())
```
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SanitizeDeviceGlobal.cpp:52]: We should probably swap the order of checks to first check for linkage and then for the name to make this `if` more efficient

Comment (jinge90) [reply llvm/lib/SYCLLowerIR/SanitizeDeviceGlobal.cpp]: Hi, @AlexeySachkov 
Done.
Thanks very much.

**Comment by jinge90** [comment]
Hi, all.
We decide not to put any non-sycl code in sycl branch, so close this PR.
Thanks very much.

**Pull request closed without merging**